eclass/rocm.eclass

```diff
@@ -0,0 +1,230 @@
+#
+# @EXAMPLE:
+# Example ebuild for ROCm library in https://github.com/ROCmSoftwarePlatform
+# which uses cmake to build and test, and depends on rocBLAS:
+# @CODE
+# ROCM_VERSION=${PV}
+# inherit cmake rocm
+# # ROCm libraries SRC_URI is usually in form of:
+# SRC_URI="https://github.com/ROCmSoftwarePlatform/${PN}/archive/rocm-${PV}.tar.gz -> ${P}.tar.gz"
+# S=${WORKDIR}/${PN}-rocm-${PV}
+# SLOT="0/$(ver_cut 1-2)"
+# IUSE="test"
+# REQUIRED_USE="${ROCM_REQUIRED_USE}"
+# RESTRICT="!test? ( test )"
+#
+# RDEPEND="
+#     dev-util/hip
+#     sci-libs/rocBLAS:${SLOT}[${ROCM_USEDEP}]
+# "
+#
+# src_configure() {
+#     # avoid sandbox violation
+#     addpredict /dev/kfd
+#     addpredict /dev/dri/
+#     local mycmakeargs=(
+#         -DAMDGPU_TARGETS="$(get_amdgpu_flags)"
+#         -DBUILD_CLIENTS_TESTS=$(usex test ON OFF)
+#     )
+#     CXX=hipcc cmake_src_configure
+# }
+#
+# src_test() {
+#     check_amdgpu
+#     # export LD_LIBRARY_PATH=<path to built lib dir> if necessary
+#     cmake_src_test # for packages using the cmake test
+#     # For packages using a standalone test binary rather than cmake test,
+#     # just execute it (or using edob)
+# }
+# @CODE
+#
+# Examples for packages depend on ROCm libraries -- a package which depends on
+# rocBLAS, uses comma separated ${HCC_AMDGPU_TARGET} to determine GPU
+# architectures, and requires ROCm version >=5.1
+# @CODE
+# ROCM_VERSION=5.1
+# inherit rocm
+# IUSE="rocm"
+# REQUIRED_USE="rocm? ( ${ROCM_REQUIRED_USE} )"
+# DEPEND="rocm? ( >=dev-util/hip-${ROCM_VERSION}
+#     >=sci-libs/rocBLAS-${ROCM_VERSION}[${ROCM_USEDEP}] )"
+#
+# src_configure() {
+#     if use rocm; then
+#         local amdgpu_flags=$(get_amdgpu_flags)
+#         export HCC_AMDGPU_TARGET=${amdgpu_flags//;/,}
+#     fi
+#     default
+# }
+# src_test() {
+#     use rocm && check_amdgpu
+#     default
+# }
+# @CODE
+
+if [[ ! ${_ROCM_ECLASS} ]]; then
+
+case ${EAPI} in
+	7|8) ;;
+	*) die "${ECLASS}: EAPI ${EAPI:-0} not supported" ;;
+esac
+
+# @ECLASS_VARIABLE: ROCM_VERSION
+# @REQUIRED
+# @PRE_INHERIT
+# @DESCRIPTION:
+# The ROCm version of current package. For ROCm libraries, it should be ${PV};
+# for other packages that depend on ROCm libraries, this can be set to match
+# the version required for ROCm libraries.
+
+# @ECLASS_VARIABLE: ALL_AMDGPU_TARGETS
+# @INTERNAL
+# @DESCRIPTION:
+# The list of USE flags corresponding to all AMDGPU targets in this ROCm
+# version. The value depends on ${ROCM_VERSION}. Architectures and devices map:
+# https://github.com/GPUOpen-Tools/device_info/blob/master/DeviceInfo.cpp
+
+# @ECLASS_VARIABLE: OFFICIAL_AMDGPU_TARGETS
+# @INTERNAL
+# @DESCRIPTION:
+# The list of USE flags corresponding to all officially supported AMDGPU
+# targets in this ROCm version, documented at
+# https://docs.amd.com/bundle/ROCm-Installation-Guide-v${ROCM_VERSION}/page/Prerequisite_Actions.html.
+# USE flag of these architectures will be default on. Depends on ${ROCM_VERSION}.
+
```

Comment:
These two do not seem to be used outside global-setter, and since they're `@INTERNAL` they're not supposed to be used outside the eclass, so I suppose it would make sense to stop setting them globally and instead make them local to that function…

Reply:
Yes, after the simplification of `get_amdgpu_flag`, it's not necessary to keep then global.